src/prefect/results.py

```diff
@@ -122,182 +121,151 @@ def _format_user_supplied_storage_key(key: str) -> str:
-            ctx.result_factory.serializer
-            if ctx and ctx.result_factory
-            else get_default_result_serializer()
-        )
-        if task.persist_result is None:
-            persist_result = (
-                ctx.result_factory.persist_result
-                if ctx and ctx.result_factory
-                else get_default_persist_setting()
-            )
+    @property
+    def serializer(self) -> Serializer:
+        # Determine the serializer
+
+        from prefect.context import FlowRunContext, TaskRunContext
+
+        flow_run_ctx = FlowRunContext.get()
+        task_run_ctx = TaskRunContext.get()
+
+        task: Optional["Task"] = getattr(task_run_ctx, "task", None)
+        flow: Optional["Flow"] = getattr(flow_run_ctx, "flow", None)
+
+        if task and task.result_serializer is not None:
+            serializer = task.result_serializer
+        elif flow and flow.result_serializer is not None:
+            serializer = flow.result_serializer
         else:
+            serializer = self._serializer
+
+        serializer = self.resolve_serializer(serializer)
+        return serializer
+
+    @property
+    def persist_result(self) -> bool:
+        # Determine if we should persist the result
+        from prefect.context import FlowRunContext, TaskRunContext
+
+        flow_run_ctx = FlowRunContext.get()
+        task_run_ctx = TaskRunContext.get()
+
+        task: Optional["Task"] = getattr(task_run_ctx, "task", None)
+        flow: Optional["Flow"] = getattr(flow_run_ctx, "flow", None)
+
+        if task and task.persist_result is not None:
             persist_result = task.persist_result
+        elif flow and flow.persist_result is not None:
+            persist_result = flow.persist_result
+        else:
+            persist_result = self._persist_result
 
-        cache_result_in_memory = task.cache_result_in_memory
-
-        return await cls.from_settings(
-            result_storage=result_storage,
-            result_serializer=result_serializer,
-            persist_result=persist_result,
-            cache_result_in_memory=cache_result_in_memory,
-            client=client,
-            storage_key_fn=(
-                partial(_format_user_supplied_storage_key, task.result_storage_key)
-                if task.result_storage_key is not None
-                else DEFAULT_STORAGE_KEY_FN
-            ),
-        )
+        return persist_result
 
-    @classmethod
-    @inject_client
-    async def from_settings(
-        cls: Type[Self],
-        result_storage: ResultStorage,
-        result_serializer: ResultSerializer,
-        persist_result: Optional[bool],
-        cache_result_in_memory: bool,
-        storage_key_fn: Callable[[], str],
-        client: "PrefectClient",
-    ) -> Self:
-        if persist_result is None:
-            persist_result = get_default_persist_setting()
+    @persist_result.setter
+    def persist_result(self, value: bool):
```

Comment:
OK so I get why these are all properties that reference hidden attributes, but I think this is where a lot of complexity is coming in; so much of this logic should (in my opinion) be owned by _creators_ of result factories, not the factory itself.

So here are two concrete proposals I have:
- make all of these basic pass through attributes from init -> attr; e.g., `self.persist_result = persist_result # as passed on instantiation`
- make _utility_ functions for each attribute that expose interfaces for determining what value to provide to the factory on init; so `result_factory = ResultFactory(persist_result=get_persist_result(task))`

I _believe_ that this will make the factory more testable, and will help co-locate business logic where it has impact (e.g., how do task result settings get made - somewhat transparently in the task engine through these utility functions).